### ISSUE TITLE
Syntaxnet fails to build

### ISSUE THREAD

**Issue by SaintNazaire** [tensorflow/models#2588]
**Issue**

Followed very exactly [Syntaxnet installation](https://github.com/tensorflow/models/tree/master/research/syntaxnet) to build syntaxnet into a docker container. Triggering the build unleashes a deluge of errors.

**System information**

- Windows 10 professional
- TensorFlow 1.3.0 installed as clean (functional) docker image from Ubuntu 16.04.3 LTS
- Docker 17.09.0-ce
- Python version 2.7
- Bazel version 0.7.0
- CUDA/cuDNN version not relevant (CPU install)
- GPU model and memory not relevant (CPU install)

**Command triggering issue**

`cd /models/research/syntaxnet && test ...  --genrule_strategy=standalone --spawn_strategy=standalone`

alternatively

`cd /models/research/syntaxnet && bazel build ... --genrule_strategy=standalone --spawn_strategy=standalone`

**Top-level directory of the model you are using**

> /models/research/syntaxnet

**Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**

Followed very exactly  [Syntaxnet installation](https://github.com/tensorflow/models/tree/master/research/syntaxnet) procedure.

Only variation I made was to clearly specify environment variables:

```
ENV PYTHON_BIN_PATH=/usr/bin/python
ENV PYTHON_LIB_PATH=/usr/lib/python2.7
```

In addition I also tried to attempt direct build instead of test and added sandboxing exception handling:

`cd /models/research/syntaxnet && bazel build ... --genrule_strategy=standalone --spawn_strategy=standalone`

**Describe the problem**

Attempt to build fails despite following exactly instruction, cf. long list of error messages below. No other step in the installation process triggers any error

**Error logs**

> Building in Docker Cloud's infrastructure...
> Cloning into '.'...
> 
(... all fine so far...)
> 
> Step 16/16 : RUN cd /models/research/syntaxnet && bazel build ... --genrule_strategy=standalone --spawn_strategy=standalone
> 
>  ---> Running in a3f7d67b4fb8
> 
... etc ... etc ... etc... etc...
> [91mINFO: From Compiling external/snappy/snappy-stubs-internal.cc [for host]:
> [0m
> [91mcc1plus: warning: command line option '-Wno-implicit-function-declaration' is valid for C/ObjC but not for C++
> cc1plus: warning: unrecognized command line option '-Wno-shift-negative-value'
... etc ... etc... etc... etc...
cf. full log attached (114 pages in Calibri 11)
[20171025 - Syntaxnet_failed_bazel_build_log.txt](https://github.com/tensorflow/models/files/1414733/20171025.-.Syntaxnet_failed_bazel_build_log.txt)

**Comment by SaintNazaire** [comment]
Looking at the logs, there is one point that seem to come back again and again:

> warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

**Comment by SaintNazaire** [comment]
Managed to narrow down the issue: either a memory maximum limit overflow or a corrupted model.

I managed to build both Syntaxnet and Dragnn individually issuing the following commands one by one:

`cd /models/research/syntaxnet && bazel build --genrule_strategy=standalone --spawn_strategy=standalone syntaxnet/... util/utf8/...`

And then upon completion of the above:

`cd /models/research/syntaxnet && bazel build --genrule_strategy=standalone --spawn_strategy=standalone dragnn/...`

**Comment by SaintNazaire** [comment]
Could we have a look at the docker image used for tensorflow/syntaxnet:latest?

**Comment by SaintNazaire** [comment]
Introspected and figured out that tensorflow/syntaxnet:latest image runs on Bazel 0.4.3 vs. Bazel 0.7.0 current version.

Ongoing test to try running with Bazel 0.4.3 instead, bear with me.

**Comment by SaintNazaire** [comment]
Test failed, attempting to build with Bazel 0.4.3 doesn't work:

`cd /models/research/syntaxnet/tensorflow && echo yes | ./configure`

> You have bazel 0.4.3 installed.
> Please upgrade your bazel installation to version 0.5.4 or higher to build TensorFlow!
> Configuration finished

`cd /models/research/syntaxnet && bazel build ... --ignore_unsupported_sandboxing`

> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/mkl/build_defs.bzl:56:18: unexpected keyword 'environ' in call to repository_rule(implementation: function, *, attrs: dict or NoneType = None, local: bool = False).
> ERROR: com.google.devtools.build.lib.packages.BuildFileContainsErrorsException: error loading package '': Extension file 'third_party/mkl/build_defs.bzl' has errors.
> INFO: Elapsed time: 0.609s

Any idea welcome.

Ongoing tests with following ideas:
- dpkg -i <.deb file>
- bazelrc file forcing --batch
- upgrading to Bazel 0.5.4

**Comment by SaintNazaire** [comment]
dpkg -i <.deb file> and bazelrc file forcing test failed

`cd /models/research/syntaxnet/tensorflow && echo yes | ./configure`

> INFO: Reading 'startup' options from /root/.bazelrc: --batch
> Extracting Bazel installation...
> You have bazel 0.4.3 installed.
> Please upgrade your bazel installation to version 0.5.4 or higher to build TensorFlow!
> Configuration finished

`cd /models/research/syntaxnet && bazel build ...`

> INFO: Reading 'startup' options from /root/.bazelrc: --batch
> WARNING: Sandboxed execution is not supported on your system and thus hermeticity of actions cannot be guaranteed. See http://bazel.build/docs/bazel-user-manual.html#sandboxing for more information. You can turn off this warning via --ignore_unsupported_sandboxing.
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/mkl/build_defs.bzl:56:18: unexpected keyword 'environ' in call to repository_rule(implementation: function, *, attrs: dict or NoneType = None, local: bool = False).
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/gpus/cuda_configure.bzl:120:22: function 'depset' does not exist.
> ERROR: com.google.devtools.build.lib.packages.BuildFileContainsErrorsException: error loading package '': Extension file 'third_party/mkl/build_defs.bzl' has errors.
> INFO: Elapsed time: 1.570s

`cd /models/research/syntaxnet && bazel build ... --ignore_unsupported_sandboxing`

> INFO: Reading 'startup' options from /root/.bazelrc: --batch
> WARNING: Sandboxed execution is not supported on your system and thus hermeticity of actions cannot be guaranteed. See http://bazel.build/docs/bazel-user-manual.html#sandboxing for more information. You can turn off this warning via --ignore_unsupported_sandboxing.
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/mkl/build_defs.bzl:56:18: unexpected keyword 'environ' in call to repository_rule(implementation: function, *, attrs: dict or NoneType = None, local: bool = False).
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/gpus/cuda_configure.bzl:120:22: function 'depset' does not exist.
> ERROR: com.google.devtools.build.lib.packages.BuildFileContainsErrorsException: error loading package '': Extension file 'third_party/mkl/build_defs.bzl' has errors.
> INFO: Elapsed time: 1.570s

`cd /models/research/syntaxnet && bazel build ... --ignore_unsupported_sandboxing`

> INFO: Reading 'startup' options from /root/.bazelrc: --batch
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/mkl/build_defs.bzl:56:18: unexpected keyword 'environ' in call to repository_rule(implementation: function, *, attrs: dict or NoneType = None, local: bool = False).
> ERROR: /root/.cache/bazel/_bazel_root/3b4c7ccb85580bc382ce4a52e9580003/external/org_tensorflow/third_party/gpus/cuda_configure.bzl:120:22: function 'depset' does not exist.
> ERROR: com.google.devtools.build.lib.packages.BuildFileContainsErrorsException: error loading package '': Extension file 'third_party/mkl/build_defs.bzl' has errors.
> INFO: Elapsed time: 2.432s

Attempting upgrading to Bazel 0.5.4

**Comment by SaintNazaire** [comment]
Unusually long install for those two components make docker hub build limits exceeded even when separating to the maximum extend each docker image build. I am forced to build locally which is not fun.

[2,855 / 2,859] 2 actions running
    Compiling external/org_tensorflow/tensorflow/core/kernels/svd_op_complex128.cc; 563s local
    Compiling external/org_tensorflow/tensorflow/core/kernels/matmul_op.cc; 557s local

**Comment by SaintNazaire** [comment]
Attempt to build with Bazel 0.5.4 failed as described in #2662
Seems Syntaxnet is no longer functional.

**Comment by SaintNazaire** [comment]
Seems the root cause of most the above is a memory issue, fixed partially as documented in #2662's closing comment.

Remains the deluge of warnings during builds (but no reported error) and several test fail, 6 for syntaxnet/... and 7 for dragnn/... with Bazel 0.5.4. Currently re-building as documented in [Syntaxnet installation](https://github.com/tensorflow/models/blob/master/research/syntaxnet/README.md) and will open dedicated issue if problem persists.

On a side not it implies that building on docker hub, even step by step will not be possible due to current limitation to 2Gb memory and therefore compilation must occur in local docker container.

**Comment by SaintNazaire** [comment]
Seems the root cause of most the above is a memory issue, fixed partially as documented in #2662's closing comment.

Remains the deluge of warnings during builds (but no reported error) and several test fail, 6 for syntaxnet/... and 7 for dragnn/... with Bazel 0.5.4. Currently re-building as documented in [Syntaxnet readme](https://github.com/tensorflow/models/blob/master/research/syntaxnet/README.md) and will open dedicated issue if problem persists.

On a side not it implies that building on docker hub, even step by step will not be possible due to current limitation to 2Gb memory and therefore compilation must occur in local docker container.
